icubam/icu_tree.py

```diff
@@ -139,16 +139,29 @@ def get_leaves(self):
     self._get_leaves(nodes)
     return nodes
 
-  def _extract_below(self, level, nodes):
-    if self.level == level:
-      nodes.append((self, self.get_leaves()))
-    else:
+  def _extract_below(self, level, nodes, keep_empty=False, max_nodes=None):
+    target = self.LEVELS.index(level)
+    curr = self.LEVELS.index(self.level)
+
+    recurse = curr < target
+    if not recurse:
+      leaves = [n for n in self.get_leaves() if (n.total > 0 or keep_empty)]
+      if max_nodes is not None and len(leaves) > max_nodes:
+        recurse = True
```

Comment:
Not sure there is an easy fix, but this (added) line is never run in tests. As it's a somewhat non trivial recursive function it wouldn't hurt to have all code paths testes.

Reply:
let me try to figure it out.

Reply:
done